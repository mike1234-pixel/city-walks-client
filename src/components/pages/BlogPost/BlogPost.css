/** @format */

.blog-post-container {
  width: 100vw;
  display: flex;
  flex-direction: column;
  align-items: center;
}

.blog-post-img {
  width: 80vw;
  margin: 20px;
}

.blog-post-content {
  width: 80vw;
}

.blog-post-comment-card {
  width: 60vw;
  margin: 20px;
  padding: 20px;
}

.add-blog-comment-form {
  padding: 50px;
}
