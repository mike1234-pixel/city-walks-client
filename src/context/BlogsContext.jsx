import { createContext, useState } from "react"

export const BlogsContext = createContext();

export const BlogsContextProvider = (props) => {

    const [blogPosts, setBlogPosts] = useState([])
    const [blogsLoading, setBlogsLoading] = useState(true)

    return (
        <BlogsContext.Provider 
            value={{
                blogPosts,
                setBlogPosts,
                blogsLoading,
                setBlogsLoading,
            }}
        >
            {props.children}
        </BlogsContext.Provider>
    )
}