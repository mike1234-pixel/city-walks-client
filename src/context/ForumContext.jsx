import { createContext, useState } from "react"
import BoardBox from "../components/pages/LoginPage/LoggedInView/Components/BoardBox/BoardBox"

export const ForumContext = createContext();

export const ForumContextProvider = (props) => {
 
    const [loadingBoards, setLoadingBoards] = useState(true)
    const [boards, setBoards] = useState([])
    const [selectedThreads, setSelectedThreads] = useState([])

    const [currentBoardName, setCurrentBoardName] = useState("")
    const [currentBoardId, setCurrentBoardId] = useState("")

    let displayBoards;

    if (!loadingBoards) {
    displayBoards = boards.map((board, index) => {
        return (
            <BoardBox boardId={board._id} name={board.name} description={board.description} index={index} key={index} />
                )
        })
        // create routes for boards with Thread as the view component
    }

    return (
        <ForumContext.Provider 
            value={{
                setBoards,
                boards, // data
                setLoadingBoards,
                loadingBoards,
                displayBoards,
                setSelectedThreads,
                selectedThreads, 
                setCurrentBoardName,
                currentBoardName,
                setCurrentBoardId,
                currentBoardId,
            }}>
            {props.children}
        </ForumContext.Provider>
    )
}

