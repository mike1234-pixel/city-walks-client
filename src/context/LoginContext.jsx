import { createContext, useState } from "react"
import axios from "axios"
import qs from "qs"
import LoginForm from "../components/pages/LoginPage/LoginForm/LoginForm";
import RegistrationForm from "../components/pages/LoginPage/RegistrationForm/RegistrationForm"
import VerificationForm from "../components/pages/LoginPage/VerificationForm/VerificationForm";
import ResetPasswordForm from "../components/pages/LoginPage/ResetPasswordForm/ResetPasswordForm";
import ForgotPasswordForm from "../components/pages/LoginPage/ForgotPasswordForm/ForgotPasswordForm";
import Boards from "../components/pages/LoginPage/LoggedInView/Boards"

export const LoginContext = createContext();

export const LoginContextProvider = (props) => {
    const [popupVisible, setPopupVisible] = useState(false)

    const [firstName, setFirstName] = useState("")
    const [lastName, setLastName] = useState("")
    const [userId, setUserId] = useState("")
    const [registrationEmail, setRegistrationEmail] = useState("")
    const [registrationPassword, setRegistrationPassword] = useState("")
  
    const [loginEmail, setLoginEmail] = useState("")
    const [loginPassword, setLoginPassword] = useState("")
  
    // save all of this in local storage
    const [loggedIn, setLoggedIn] = useState(false)
    const [userFirstName, setUserFirstName] = useState("")
    const [userLastName, setUserLastName] = useState("")
  
    const [verificationEmail, setVerificationEmail] = useState("")

    const [resetPasswordEmail, setResetPasswordEmail] = useState("")
    const [resetPasswordOldPassword, setResetPasswordOldPassword] = useState("")
    const [resetPasswordNewPassword, setResetPasswordNewPassword] = useState("")
    const [resetPasswordConfirmNewPassword, setResetPasswordConfirmNewPassword] = useState("")

    const [forgotPasswordEmail, setForgotPasswordEmail] = useState("")

    const handlePopup = () => {
      setPopupVisible(false)
      localStorage.setItem("popupVisible", false)
    }

    // set display form on LoginPage
    const [form, setForm] = useState("loggedOutView");

    let displayForm = <Boards/>;
    if (form === "boards") {
      displayForm = <Boards/>;
    } else if (form === "loginForm") {
      displayForm = <LoginForm />;
    } else if (form === "registrationForm") {
      displayForm = <RegistrationForm />;
    } else if (form === "verificationForm") {
      displayForm = <VerificationForm />;
    } else if (form === "resetPasswordForm") {
      displayForm = <ResetPasswordForm />;
    } else if (form === "forgotPasswordForm") {
      displayForm = <ForgotPasswordForm />;
    }

    const handleChangeRegistration = (event) => {
        switch(event.target.name) {
            case "registration-fname":
              setFirstName(event.target.value)
              break;
            case "registration-lname":
              setLastName(event.target.value)
              break;
            case "registration-email":
              setRegistrationEmail(event.target.value)
              break;
            case "registration-password":
              setRegistrationPassword(event.target.value)
              break;
          } 
    }
  
    const handleSubmitRegistration = (event) => {
      console.log("handle submit triggered")
        event.preventDefault()
  
        const payload = {
            fname: firstName,
            lname: lastName,
            email: registrationEmail,
            password: registrationPassword,
          };
  
    axios
        .post("http://localhost:5000/register-user", qs.stringify(payload))
        .then((res, err) => {
          if (err) {
            console.log(err);
          } else if (res.data === "An account with this email already exists.") {
            alert("An account with this email already exists.")
          } else if (res.data === "We have sent you an email. Please verify your account by clicking the link in the mail.") {
            alert("We have sent you an email. Please verify your account by clicking the link in the email. (This code expires after 10 minutes)")
           } else {
            alert("Registration Complete")
            setLoggedIn(true)
            setFirstName("")
            setLastName("")
            setRegistrationEmail("")
            setRegistrationPassword("")
            setUserFirstName(res.data.fname)
            setUserLastName(res.data.lname)
            window.scrollTo(0, 0)
          }
        });
    }
  
    const handleChangeLogin = (event) => {
      switch(event.target.name) {
          case "login-email":
            setLoginEmail(event.target.value)
            break;
          case "login-password":
            setLoginPassword(event.target.value)
            break;
        } 
  }
  
  const handleSubmitLogin = (event) => {
    console.log("handle submit login triggered")
      event.preventDefault()
  
      const payload = {
          email: loginEmail,
          password: loginPassword
        };
  
  axios
      .post("http://localhost:5000/login-user", qs.stringify(payload))
      .then((res, err) => {
        if (err) {
          console.log(err);
        } else if (res.data === "Your account exists but is not activated. Please click 'verify account' for email verification.") {
          alert("Your account exists but is not activated. Please click 'verify account' for email verification.")
        } else if (res.data === "unsuccessful login attempt") {
          alert("Unsuccessful Login Attempt. Please Try Again.")
        } else {
          alert("Login Complete")
          setLoggedIn(true)
          setLoginEmail("")
          setLoginPassword("")
          setForm("")
          setUserId(res.data._id)
          setUserFirstName(res.data.fname)
          setUserLastName(res.data.lname)
          localStorage.setItem("loggedIn", true)
          localStorage.setItem("userId", res.data._id)
          localStorage.setItem("userFirstName", res.data.fname)
          window.scrollTo(0, 0)
        }
      });
  }

  // resend verification email

  const handleChangeVerification = (event) => {
    switch(event.target.name) {
        case "verification-email":
          setVerificationEmail(event.target.value)
          break;
      }   
  }

const handleSubmitVerification = (event) => {
    console.log("handle submit verification")
    event.preventDefault()

    const payload = {
        email: verificationEmail,
      };

axios
    .post("http://localhost:5000/reverify-user", qs.stringify(payload))
    .then((res, err) => {
      if (err) {
        console.log(err);
      } else {
        alert("Verification email sent. Check your inbox.")
        setVerificationEmail("")
        window.scrollTo(0, 0)
      }
    })
}

  // reset password

  const handleChangeResetPassword = (event) => {
    switch(event.target.name) {
        case "reset-email":
          setResetPasswordEmail(event.target.value)
          break;
        case "old-password":
          setResetPasswordOldPassword(event.target.value)
          break;
        case "new-password":
          setResetPasswordNewPassword(event.target.value)
          break;
        case "confirm-new-password":
          setResetPasswordConfirmNewPassword(event.target.value)
          break;
      }   
  }

const handleSubmitResetPassword = (event) => {
    console.log("handle submit reset password")
    event.preventDefault()

    const payload = {
        email: resetPasswordEmail,
        oldPassword: resetPasswordOldPassword,
        newPassword: resetPasswordNewPassword,
      };

      if (resetPasswordNewPassword === resetPasswordConfirmNewPassword) {
        axios
        .post("http://localhost:5000/reset-password-with-old-password", qs.stringify(payload))
        .then((res, err) => {
          if (err) {
            console.log(err);
          } else if (res.data === "user not found") {
            alert("We could not find your account. Please try again.")
          } else if (res.data === "old password does not match password in the database") {
            alert("The password you entered does not match the account. Please try again or use the Forgot Password option.")
          } else if (res.data === "password successfully updated") {
            alert("Password reset. You can now login.")
            setResetPasswordEmail("")
            setResetPasswordOldPassword("")
            setResetPasswordNewPassword("")
            setResetPasswordConfirmNewPassword("")
            window.scrollTo(0, 0)
          }
        })
      } else {
        alert("Passwords don't match. Please try again.")
      }

}

  // forgot password

  const handleChangeForgotPassword = (event) => {
    switch(event.target.name) {
        case "forgot-password-email":
          setForgotPasswordEmail(event.target.value)
          break;
      }   
  }

const handleSubmitForgotPassword = (event) => {
    console.log("handle submit forgot password")
    event.preventDefault()

    const payload = {
        email: forgotPasswordEmail,
      };

axios
    .post("http://localhost:5000/forgot-password", qs.stringify(payload))
    .then((res, err) => {
      if (err) {
        console.log(err);
      } else {
        alert("We have sent you an email. Please click the click in your email to reset your password")
        setForgotPasswordEmail("")
        window.scrollTo(0, 0)
      }
    })
}
  
  const logOut = () => {
    localStorage.clear()
    localStorage.setItem("popupVisible", false)
    setLoggedIn(false)
    setUserId("")
    setUserFirstName("")
    setUserLastName("")
    setForm("boards")
  }

    return (
        <LoginContext.Provider 
            value={{
              popupVisible,
              setPopupVisible,
              handlePopup,
               // display form
               displayForm,
               setForm,
               // logged in state 
               loggedIn,
               // registration state and functions
                firstName,
                lastName,
                registrationEmail,
                registrationPassword,
                handleChangeRegistration,
                handleSubmitRegistration,
                // login state and functions
                loginEmail,
                loginPassword,
                handleChangeLogin: handleChangeLogin,
                handleSubmitLogin: handleSubmitLogin,
                setLoggedIn,
                setUserFirstName,
                setUserId,
                // logged in user data
                userFirstName,
                userId,
                userLastName,
                // resend verification email
                verificationEmail,
                handleChangeVerification,
                handleSubmitVerification,
                // reset password
                resetPasswordEmail,
                resetPasswordOldPassword,
                resetPasswordNewPassword,
                resetPasswordConfirmNewPassword,
                handleChangeResetPassword,
                handleSubmitResetPassword,
                // forgot password
                forgotPasswordEmail,
                handleChangeForgotPassword,
                handleSubmitForgotPassword,
                // logout function
                logOut
            }}
        >
            {props.children}
        </LoginContext.Provider>
    )
}

// // setter
// localStorage.setItem('myData', data);
 
// // getter
// localStorage.getItem('myData');
 
// // remove
// localStorage.removeItem('myData');
 
// // remove all
// localStorage.clear();

