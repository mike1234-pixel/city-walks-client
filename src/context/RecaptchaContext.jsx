import { createContext, useEffect } from "react"

export const RecaptchaContext = createContext();

export const RecaptchaContextProvider = (props) => {
    
  const siteKey = "6LdmxiUaAAAAAIYySt3c8XvwOMokTQ_SW2cYkvMw";
 
  // dynamically load the javascript api using the site key
    useEffect(() => {
      const loadScriptByURL = (id, url, callback) => {
        const isScriptExist = document.getElementById(id);
    
        if (!isScriptExist) {
          var script = document.createElement("script");
          script.type = "text/javascript";
          script.src = url;
          script.id = id;
          script.onload = function () {
            if (callback) callback();
          };
          document.body.appendChild(script);
        }
    
        if (isScriptExist && callback) callback();
      }
    
      // load the script by passing the URL
      loadScriptByURL("recaptcha-key", `https://www.google.com/recaptcha/api.js?render=${siteKey}`, () => {
        console.log("recaptcha script loaded!");
      });
    }, []);

    return (
        <RecaptchaContext.Provider 
            value={{
            siteKey
            }}>
            {props.children}
        </RecaptchaContext.Provider>
    )
}