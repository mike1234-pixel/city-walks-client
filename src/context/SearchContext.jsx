import { createContext, useState } from "react"

export const SearchContext = createContext();

export const SearchContextProvider = (props) => {
    const [searchValue, setSearchValue] = useState("")
    const [redirect, setRedirect] = useState(false)
  
    const handleChangeSearch = (e) => {
      setSearchValue(e.target.value)
    }
  
    const handleSubmitSearch = (e) => {
      e.preventDefault();
      setRedirect(true);
    }
  
    const handleClickSearch = (city) => {
      setSearchValue(city)
      setRedirect(true);
    }

    return (
        <SearchContext.Provider 
            value={{
            handleChangeSearch, 
            handleSubmitSearch,  
            redirect, 
            setRedirect,
            handleClickSearch, 
            searchValue,
            setSearchValue
            }}
        >
            {props.children}
        </SearchContext.Provider>
    )
}