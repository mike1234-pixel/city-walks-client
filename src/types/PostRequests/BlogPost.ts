export default interface BlogPost {
    title: string;
    subtitle: string;
    content: string;
    img: string;
    author: string;
}