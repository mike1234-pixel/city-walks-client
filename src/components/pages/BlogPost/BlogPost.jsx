import { useEffect, useContext, useState } from "react"
import { BlogsContext } from "../../../context/BlogsContext"
import PopUp from "../../PopUp/PopUp"
import { LoginContext } from "../../../context/LoginContext"
import toTitleCase from "../../../functions/toTitleCase"
import { MDBInput, MDBBtn, MDBIcon, MDBCard, MDBCardTitle, MDBCardText } from "mdbreact"
import axios from "axios"
import qs from "qs"
import marked from "marked";
import './BlogPost.css'

const BlogPost = ({match}) => {

    // useEffect(() => {
    //     window.scrollTo(0, 0);
    //   });

      const blogTitle = toTitleCase(match.url.replace("/blog/", "").replace(/-/g, " "))

      const { blogPosts, blogsLoading } = useContext(BlogsContext)
      const { loggedIn, userFirstName, userId, popupVisible } = useContext(LoginContext)

      const [comment, setComment] = useState("")
      const [currentBlogTitle, setCurrentBlogTitle] = useState("")

      const handleChange = (event) => {
        setComment(event.target.value)
      }

      const handleSubmit = (event) => {
        event.preventDefault()

        let payload = {
          currentBlogTitle: blogTitle,
          comment: comment,
          userFirstName: userFirstName,
          userId: userId
        };

      axios
      .post("http://localhost:5000/add-blog-comment", qs.stringify(payload))
      .then((res, err) => {
        if (err) {
          console.log(err);
        } else if (res.data === "comment submitted") {
          console.log("comment submitted")
        } else {
          console.log("comment not submitted")
        }
      });

      alert("Comment Submitted")
      setComment("")
      window.location.reload()
      }

      const handleDeleteComment = (commentId) => {

        let payload = {
          currentBlogTitle: blogTitle,
          commentId: commentId,
        };
  
        axios
        .delete("http://localhost:5000/delete-blog-comment", { data: payload })
        .then((res, err) => {
          if (err) {
            console.log(err);
          } else {
            alert("comment deleted.")
            window.location.reload()
          }
        });
      }

      let post = "loading"

      if (!blogsLoading) {

      let selectedBlogPost = blogPosts.filter((post) => post.title === blogTitle)
      selectedBlogPost = selectedBlogPost[0]
    
      const createMarkup = (markup) => {
        return {__html: marked(markup, {breaks: true})}
      }

      if (selectedBlogPost === undefined) {
        post = "walk not found"
      } else {
        post = 
        <div className="blog-post-container">
            <h1 className="page-heading">{selectedBlogPost.title}</h1>
            <h2 className="blog-subtitle page-heading">{selectedBlogPost.subtitle}</h2>
            <img className="blog-post-img" src={selectedBlogPost.img}/>
            <div className="blog-post-content" dangerouslySetInnerHTML={createMarkup(selectedBlogPost.content)}></div>
            <p>{selectedBlogPost.submittedOn.replace('T', ' ').substring(0, 19)}</p>
            {selectedBlogPost.comments.map((comment) => {
              return (
                <MDBCard className="blog-post-comment-card" key={comment._id}>
                  <MDBCardTitle>{comment.userFirstName} commented:</MDBCardTitle>
                  <MDBCardText>{comment.comment}</MDBCardText>
                  <MDBCardText>{comment.submittedOn.replace('T', ' ').substring(0, 19)}</MDBCardText>
                  {userId === comment.userId && <MDBBtn className="blog-post-comment-card-btn" onClick={() => handleDeleteComment(comment._id)}>Delete Comment</MDBBtn>}
                </MDBCard>
              )
            })}
        </div>
      }
    }

    return (
        <div>
            {popupVisible && <PopUp/>}
            <div>{post}</div>
            {loggedIn &&
              <form onSubmit={handleSubmit} className="add-blog-comment-form">
                <MDBInput type="textarea" rows="5" name="comment" id="comment" value={comment} label="comment" onChange={handleChange} required/>
                <MDBBtn outline color="elegant" type="submit">
                    Add Comment <MDBIcon icon="plus"/>
                </MDBBtn>
              </form>
            }
        </div>
    )
}

export default BlogPost