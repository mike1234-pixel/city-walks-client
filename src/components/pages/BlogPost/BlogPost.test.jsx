import React from 'react'
import renderer from 'react-test-renderer'
import { BlogsContextProvider } from "../../../context/BlogsContext"
import { LoginContextProvider } from "../../../context/LoginContext"
import BlogPost from './BlogPost'

// snapshot test
it('Blog Post renders correctly', () => {  

  const tree = renderer
    .create(
        <LoginContextProvider>
            <BlogsContextProvider>
                <BlogPost match={{params: {id: 1}, isExact: true, path: "", url: "/blog/blog-title-1"}}/>
            </BlogsContextProvider>
        </LoginContextProvider>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});