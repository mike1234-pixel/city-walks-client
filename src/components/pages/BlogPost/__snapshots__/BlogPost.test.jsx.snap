// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Blog Post renders correctly 1`] = `
<div>
  <div>
    loading
  </div>
</div>
`;
